spacy/visualization.py

```diff
@@ -0,0 +1,601 @@
+                continue
+            first_index_in_relation = min(working_token_index, head_token_index)
+            second_index_in_relation = max(working_token_index, head_token_index)
+            char_horizontal_line_length = (
+                2 * horizontal_line_lengths[working_token_index]
+            )
+
+            # Draw the corners of the relation
+            char_matrix[first_index_in_relation][char_horizontal_line_length - 1] |= (
+                HALF_HORIZONTAL_LINE + LOWER_HALF_VERTICAL_LINE
+            )
+            char_matrix[second_index_in_relation][char_horizontal_line_length - 1] |= (
+                HALF_HORIZONTAL_LINE + UPPER_HALF_VERTICAL_LINE
+            )
+
+            # Draw the horizontal line for the governing token
+            for working_horizontal_position in range(char_horizontal_line_length - 1):
+                if (
+                    char_matrix[head_token_index][working_horizontal_position]
+                    != FULL_VERTICAL_LINE
+                ):
+                    char_matrix[head_token_index][
+                        working_horizontal_position
+                    ] |= FULL_HORIZONTAL_LINE
+
+            # Draw the vertical line for the relation
+            for working_vertical_position in range(
+                first_index_in_relation + 1, second_index_in_relation
+            ):
+                if (
+                    char_matrix[working_vertical_position][
+                        char_horizontal_line_length - 1
+                    ]
+                    != FULL_HORIZONTAL_LINE
+                ):
+                    char_matrix[working_vertical_position][
+                        char_horizontal_line_length - 1
+                    ] |= FULL_VERTICAL_LINE
+        for working_token_index in (
+            i for i in range(sent.end - sent.start) if heads[i] is not None
+        ):
+            for working_horizontal_position in range(
+                2 * horizontal_line_lengths[working_token_index] - 2, -1, -1
+            ):
+                if (
+                    (
+                        char_matrix[working_token_index][working_horizontal_position]
+                        == FULL_VERTICAL_LINE
+                    )
+                    and working_horizontal_position > 1
+                    and char_matrix[working_token_index][
+                        working_horizontal_position - 2
+                    ]
+                    == SPACE
+                ):
+                    # Cross over the existing vertical line, which is owing to a non-projective tree
+                    continue
+                if (
+                    char_matrix[working_token_index][working_horizontal_position]
+                    != SPACE
+                ):
+                    # Draw the arrowhead to the right of what is already there
+                    char_matrix[working_token_index][
+                        working_horizontal_position + 1
```

Comment:
Nitpick, but I wonder if we can come up with some shorter variables names across this module? It seems like we're causing a lot of unnecessary line breaks during formatting because the lines get too long. For instance, I'd be fine with shortening `horizontal_position` to `horiz_pos` etc.